[ML] Datafeed config CRUD operations

davidkyle opened the pull request:
Similar to #32738 this time for the datafeed configuration document.

I factored out the expanded Ids logic from `JobConfigProvider` into a new class and have reused here

elasticmachine commented:
Pinging @elastic/ml-core

davidkyle commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfigTests.java:
```diff
@@ -167,6 +180,33 @@ public void testFutureMetadataParse() throws IOException {
         assertNotNull(DatafeedConfig.LENIENT_PARSER.apply(parser, null).build());
     }
 
+    public void testToXContentForClusterState() throws IOException {
```
I wanted to add a test to check the datafeed type isn't written when we are writing to cluster state i.e. this piece of code in DatafeedConfig::toXContent
```
if (params.paramAsBoolean(ToXContentParams.FOR_CLUSTER_STATE, false) == false) {
            builder.field(CONFIG_TYPE.getPreferredName(), TYPE);
        }
``` 

But it's hard to test so I added a check for the headers which similarly should only be persisted in certain circumstances

benwtrent replied:
Why not the following something similar to the following? It may work to simple directly check the serialized data string.

```
String stringToCheck = "\""+DatafeedConfig.CONFIG_TYPE.getPreferredName()+"\":\""+DatafeedConfig.TYPE+"\"";
assertThat(forClusterstateXContent.utf8ToString(), containsString(stringToCheck));
...
assertThat(forClusterstateXContent.utf8ToString(), not(containsString(stringToCheck)));
```

davidkyle replied:
Yes that works, great idea I never thought of it.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java:
```diff
@@ -304,14 +320,14 @@ public void writeTo(StreamOutput out) throws IOException {
+        // Config type field is added for the migration to index documents
+        // and isn't needed in cluster state configs. Not writing the field
+        // protects against BWC issues.
+        if (params.paramAsBoolean(ToXContentParams.FOR_CLUSTER_STATE, false) == false) {
```
Originally this `ToXContentParams.FOR_CLUSTER_STATE` parameter was used to avoid showing the user internal implementation details.  Based on the original intent the parameter should probably have been named `ToXContentParams.FOR_INTERNAL_STORAGE`.

There is also another constant, `MlMetaIndex.INCLUDE_TYPE_KEY`, which is used to avoid showing the end user the metadata document type in REST responses.  I think we should also use that for config documents.  So maybe move `INCLUDE_TYPE_KEY` into `ToXContentParams` so it can be reused here.

Then the logic here will be to include the type neither for cluster state nor for REST responses.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java:
```diff
@@ -336,6 +352,7 @@ public XContentBuilder doXContentBody(XContentBuilder builder, Params params) th
         if (headers.isEmpty() == false && params.paramAsBoolean(ToXContentParams.FOR_CLUSTER_STATE, false) == true) {
             builder.field(HEADERS.getPreferredName(), headers);
```
We'll need to store the headers in the indexed document.

droberts195 commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfigTests.java:
```diff
@@ -167,6 +180,41 @@ public void testFutureMetadataParse() throws IOException {
+        String configTypeEntry = "\""+DatafeedConfig.CONFIG_TYPE.getPreferredName()+"\":\""+DatafeedConfig.TYPE+"\"";
+        assertThat(forClusterstateXContent.utf8ToString(), not(containsString(configTypeEntry)));
+
+        // headers are not written without the FOR_CLUSTER_STATE param
```
They should be written for all internal storage, whether to index or cluster state.  They should not be written in REST responses.

davidkyle commented:
I renamed FOR_CLUSTER_STATE and moved INCLUDE_TYPE_KEY. 

The lenient cluster state parser is in 5.6 so I don't need the check on writing the new `DatafeedConfig.CONFIG_TYPE.getPreferredName() : DatafeedConfig.TYPE` field as even in rolling upgrade the new field is tolerated

davidkyle commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfigTests.java:
```diff
@@ -100,7 +113,7 @@ public static DatafeedConfig createRandomizedDatafeedConfig(String jobId, long b
                 builder.setFrequency(TimeValue.timeValueSeconds(randomIntBetween(1, 1_000_000)));
             } else {
-                builder.setFrequency(TimeValue.timeValueMillis(randomIntBetween(1, 5) * aggHistogramInterval));
+                builder.setFrequency(TimeValue.timeValueSeconds(randomIntBetween(1, 5) * aggHistogramInterval));
```
This test failed with the error:
```
java.lang.IllegalArgumentException: frequency has to be a multiple of 1s; actual was '900ms'
...
   >    at org.elasticsearch.xpack.core.ml.utils.time.TimeUtils.checkMultiple(TimeUtils.java:101)
   >    at org.elasticsearch.xpack.core.ml.utils.time.TimeUtils.checkPositiveMultiple(TimeUtils.java:74)
```

It's clear that `TimeValue.timeValueMillis(randomIntBetween(1, 5) * aggHistogramInterval)` is unlikely to return a whole number of seconds but I'm amazed it has not happened before.

Reproduce command was
```
/gradlew :x-pack:plugin:core:test -Dtests.seed=7D58D672CDD2EF69 -Dtests.class=org.elasticsearch.xpack.core.ml.datafeed.DatafeedConfigTests -Dtests.method="testToXContentForInternalStorage" -Dtests.security.manager=true -Dtests.locale=ln -Dtests.timezone=America/Rankin_Inlet -Dcompiler.java=10 -Druntime.java=10
```

I'll make the same change in `...protocol...DatafeedConfigTests`

davidkyle replied:
Actually there is no validation in `protocol..DatafeedConfigTests` so not an issue

davidkyle commented:
Jobs have the field `job_type : anomaly_detector` (job_type is always anomaly_detector), that field is serialised used in the queries
